### PR TITLE
Compile on-run-(start|end) hooks to file

### PR DESCRIPTION
**Pull request by drewbanin**
@cmcarthur i'm not super happy with this yet -- going to try to clean up a little in the AM

### CODE REVIEW

File: dbt/compilation.py [hunk 1]
```diff
@@ -351,11 +355,13 @@ def link_graph(self, linker, flat_graph):
         linked_graph = {
             'nodes': {},
             'macros': flat_graph.get('macros'),
+            'operations': flat_graph.get('operations'),
         }
 
-        for name, node in flat_graph.get('nodes').items():
-            self.link_node(linker, node, flat_graph)
-            linked_graph['nodes'][name] = node
+        for node_type in ['nodes', 'operations']:
```
Comment (cmcarthur) [inline dbt/compilation.py]: curious why you put operations in a separate subgraph

File: dbt/parser.py [hunk 2]
```diff
@@ -331,6 +331,52 @@ def load_and_parse_sql(package_name, root_project, all_projects, root_dir,
     return parse_sql_nodes(result, root_project, all_projects, tags)
```
Comment (cmcarthur) [inline dbt/parser.py]: looks like this file was `chmod +x`ed -- can you undo that

File: dbt/parser.py [hunk 3]
```diff
@@ -331,6 +331,52 @@ def load_and_parse_sql(package_name, root_project, all_projects, root_dir,
...
+        result.append({
+            'name': hook_type,
+            'root_path': "{}/dbt_project.yml".format(project_name),
+            'resource_type': NodeType.Operation,
+            'path': hook_path,
+            'package_name': project_name,
+            'raw_sql': hooks
+        })
+
+    return parse_sql_nodes(result, root_project, all_projects, tags={hook_type})
```
Comment (cmcarthur) [inline dbt/parser.py]: looks great

File: dbt/runner.py [hunk 4]
```diff
@@ -755,18 +734,35 @@ def get_idx(node):
...
+        for ancestor in all_ancestors:
+            if ancestor not in flat_graph['nodes']:
+                continue
+            ancestor_node = flat_graph['nodes'][ancestor]
+            is_model = ancestor_node.get('resource_type') == NodeType.Model
+            is_ephemeral = get_materialization(ancestor_node) == 'ephemeral'
+            if is_model and is_ephemeral:
+                res.append(ancestor)
+
+        return set(res)
```
Comment (cmcarthur) [inline dbt/runner.py]: 👍

File: dbt/runner.py [hunk 5]
```diff
@@ -874,15 +870,17 @@ def compile_models(self, include_spec, exclude_spec):
...
         ]
 
         return self.run_types_from_graph(include_spec,
                                          exclude_spec,
                                          resource_types=resource_types,
                                          tags=set(),
                                          should_run_hooks=False,
-                                         should_execute=False)
+                                         should_execute=False,
+                                         flatten_graph=True)
```
Comment (cmcarthur) [inline dbt/runner.py]: why flatten the graph here?

File: dbt/compilation.py (hunk 1 above)
Comment (drewbanin) [inline dbt/compilation.py]: yeah, i think you're right -- these can in the `nodes` graph... let me investigate

File: dbt/runner.py (hunk 5 above)
Comment (drewbanin) [inline dbt/runner.py]: debugging! good catch

**Comment by drewbanin** [comment]
this fixes https://github.com/fishtown-analytics/dbt/issues/403

File: dbt/utils.py [hunk 6]
```diff
@@ -31,6 +31,12 @@ class NodeType(object):
     Operation = 'operation'
 
 
+class RunHookTypes:
```
Comment (cmcarthur) [inline dbt/utils.py]: `RunHookType`

Comment (drewbanin) [inline dbt/utils.py]: reasonable

**Pull request merged**